### DISCUSSION TITLE
방어적 복사와 불변 컬렉션은 언제 사용하면 좋나요?

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#50] [Clean Code]
입력 = 방어적 복사
응답 = 성능을 고려해서 불변 컬렉션 고고?

/**
     * 불변 컬렉션을 사용하여 객체의 상태를 변경할 수 있는 위험을 방지하는 방법입니다.
     * 응답하는 컬렉션을 불변 컬렉션으로 만들어 객체의 상태를 변경할 수 있는 위험을 방지할 수 있습니다. 
     * 입력을 받는 컬렉션을 불변 컬렉션을 만드는 것은 그대로 외부에서 조작할 수 있는 위험이 존재합니다.
     * 따라서 입력받는 컬렉션은 방어적 복사로, 응답하는 컬렉션은 불변 컬렉션으로 만드는 것이 좋습니다.
     */

**Comment by le2sky** [comment]
실수를 피하는 코드에 대한 내용이므로, 클린 코드로 분류 해봤어용

**Comment by jimi567** [comment]
방어적 복사와 불변 컬렉션은 데이터의 무결성을 보장하기 위한 상황에서 고려할 수 있는 방법 입니다.
예를 들면,
메소드나 클래스를 설계할 때 매개변수로 전달되는 컬렉션의 상태를 보장해주어야 할 때, 사용합니다. 방어적 복사를 통해 외부에서 잔달되는 컬렉션의 원본을 보호하고, 불변 컬렉션을 사용해 클라이언트 코드에서 컬렉션을 변경하지 못하도록 할 수 있습니다.

또한, 방어적 복사를 사용하면 컬렉션을 전달할 때 실수로 참조를 유지하고 있는 경우에 발생할 수 있는 메모리 누수를 방지할 수 있습니다.
함수형 프로그래밍에서도 불변 컬렉션을 사용하면, 함수형 프로그래밍의 원칙을 준수하며 side - effect을 최소화할 수 있습니다.

방어적 복사를 사용하여 스레드 간의 데이터 공유를 제어하거나, 불변 컬렉션을 사용하여 변경이 불가능한 상태로 만드는 등으로 멀티 스레드 환경에서 안전성을 보장할 때도 방어적 복사나 불변 컬렉션을 고려할 수 있습니다.
(이든 테코톡)

**Comment by nayonsoso** [comment]
※ 질문에서의 방어적 복사는 `new ArrayList`로 / 불변 컬렉션은 `Collections.unmodifiableList` 로 해석했습니다 ※

불변 컬렉션은 '변경 불가능한 뷰'를 제공합니다.
따라서 외부에서 컬렉션에 대해 수정을 하려 하면 예외가 발생합니다.
하지만, 원본 컬렉션이 변경된다면 그 변경이 뷰에도 반영이 됩니다.
저는 이런 특성이 getter와 어울린다고 생각합니다.

방어적 복사는 기존 컬렉션과의 참조를 끊고, 새로운 컬렉션을 만듭니다.
때문에 기존 컬렉션을 수정하고 싶을 때 사용할 수 있습니다.
하지만 컬렉션을 방어적 복사해도, 내부 요소들은 기존 요소들을 참조하기 때문에
기존 컬렉션의 요소가 변하면 새로운 컬렉션도 영향을 받습니다.
이를 방지하기 위해서 요소 자체를 불변으로 만들 수 있습니다.
또는 내부 인스턴스와 완전히 같은 값을 갖는 새로은 인스턴스를 만들어주는 방법도 있는데요,
이 방법은 객체 생성의 비용이 크다고 생각해서, 객체를 불변으로 만드는 방법을 택할 것 같습니다.

산초 정리 : https://github.com/nayonsoso/WIL/blob/main/level1/2-ladder-5.md

**Comment by GIVEN53** [comment]
방복
참조 끊음
새로운 컬렉션
무결성 보장

불컬
변경 불가능한 view
외부에서 수정 못하게
